base/dict.jl

```diff
@@ -104,12 +104,8 @@ mutable struct Dict{K,V} <: Associative{K,V}
         new(zeros(UInt8,n), Array{K,1}(n), Array{V,1}(n), 0, 0, 0, 1, 0)
     end
     function Dict{K,V}(d::Dict{K,V}) where V where K
-        if d.ndel > 0
-            rehash!(d)
-        end
-        @assert d.ndel == 0
-        new(copy(d.slots), copy(d.keys), copy(d.vals), 0, d.count, d.age, d.idxfloor,
-            d.maxprobe)
+        new(copy(d.slots), copy(d.keys), copy(d.vals), d.ndel, d.count, d.age,
+            d.idxfloor, d.maxprobe)
```

Comment:
Should we do `dnew = new(...)` and then `rehash!(dnew)` if `d.ndel > 0`?

Reply:
This is exactly what I have considered (as given in the comment to the PR). I have decided not to add it as it introduces a performance penalty for `Dict` creation. Probably @JeffBezanson can provide the guidance here.
Another approach would be to test `d.ndel > "some value greater than 0"` (i.e. if we have a lot of deleted entries we do `rehash!`, and if there are only a few we leave it without rehashing).

Reply:
As a second thought - my recommendation would be to remove `rehash!` from the constructor. Its main benefit is that later we do not have to `rehash!` both dictionaries twice. But this is something that we cannot avoid anyway.
If we believe that being able to a clean dictionary is an important feature I would rather recommend to:
* rename `rehash!` to `_rehash!`;
* expose a new `rehash!` as an exported function and define it for `Dict` and `Set`. This new `rehash!` should perform an additional check if the size is not decreased (as `_rehash!` is designed for growing only).